### PR TITLE
Calculate skip indexes on vertical stage of merge

### PR DESCRIPTION
**Pull request by CurtizJ**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Optimized memory usage of vertical merges for tables with high number of skip indexes.

<details>
    <summary>CI Settings</summary>

**NOTE:** If your merge the PR with modified CI you **MUST KNOW** what you are doing
**NOTE:** Checked options will be applied if set before CI RunConfig/PrepareRunConfig step
- [ ]  Allow: Integration Tests
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Unit tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All with aarch64
- [ ]  Allow: All with ASAN
- [ ]  Allow: All with TSAN
- [ ]  Allow: All with Analyzer
- [ ]  Allow: All with Azure
- [ ]  Allow: Add your option here
---
- [ ]  Exclude: Fast test
- [ ]  Exclude: Integration Tests
- [ ]  Exclude: Stateless tests
- [ ]  Exclude: Stateful tests
- [ ]  Exclude: Performance tests
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN
- [ ]  Exclude: All with MSAN
- [ ]  Exclude: All with UBSAN
- [ ]  Exclude: All with Coverage
- [ ]  Exclude: All with Aarch64
---
- [ ]  do not test (only style check)
- [ ]  disable merge-commit (no merge from master before tests)
- [ ]  disable CI cache (job reuse)
- [ ]  allow: batch 1 for multi-batch jobs
- [ ]  allow: batch 2
- [ ]  allow: batch 3
- [ ]  allow: batch 4, 5 and 6
</details>

### CODE REVIEW

**Comment by robot-ch-test-poll1** [comment]
*This is an automated comment for commit 1175c3d27ed157d68246db5b46a4dac638e2e474 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/64580/1175c3d27ed157d68246db5b46a4dac638e2e474/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>A Sync</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>⏳ pending</td></tr>
<tbody>
</table>
<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

**Comment by CurtizJ** [comment]
Comparison of memory usage on table with about 250 columns and 22 skip index. The contains about 5Gb (both compressed and uncompressed) of random data.

On local disk:
```
   ┌────event_time_microseconds─┬─merge_algorithm─┬─length(merged_from)─┬─read_rows─┬─duration_ms─┬─memory_usage─┐
1. │ 2024-05-30 18:02:00.868771 │ Vertical        │                  26 │   2998892 │       61818 │ 538.55 MiB   │ --before
2. │ 2024-05-30 18:02:55.451000 │ Vertical        │                   2 │   3000000 │       54581 │ 202.08 MiB   │

3. │ 2024-05-30 18:05:20.286644 │ Vertical        │                  26 │   2998892 │       59163 │ 83.79 MiB    │ -- after
4. │ 2024-05-30 18:06:10.338851 │ Vertical        │                   2 │   3000000 │       50051 │ 41.58 MiB    │
   └────────────────────────────┴─────────────────┴─────────────────────┴───────────┴─────────────┴──────────────┘
```

On S3:
```  
    ┌────event_time_microseconds─┬─merge_algorithm─┬─length(merged_from)─┬─read_rows─┬─duration_ms─┬─memory_usage─┐
 1. │ 2024-05-30 18:15:51.846946 │ Vertical        │                  26 │   2998892 │      102120 │ 5.60 GiB     │ -- before
 2. │ 2024-05-30 18:16:55.267398 │ Vertical        │                   2 │   3000000 │       63419 │ 1.57 GiB     │

 3. │ 2024-05-30 18:33:26.065638 │ Vertical        │                  26 │   2998892 │       67208 │ 5.23 GiB     │ -- after
 4. │ 2024-05-30 18:34:21.787561 │ Vertical        │                   2 │   3000000 │       55721 │ 614.17 MiB   │
    └────────────────────────────┴─────────────────┴─────────────────────┴───────────┴─────────────┴──────────────┘
```

For some reason merge on S3 uses unbelievably huge amount of memory for vertical merge for table that have only 5Gb of data. Need to investigate this, but it's out of scope of this PR.

File: src/Storages/MergeTree/MergeTask.cpp [hunk 1]
```diff
@@ -969,14 +997,16 @@ void MergeTask::ExecuteAndFinalizeHorizontalPart::createMergedStream()
...
+
     for (const auto & part : global_ctx->future_part->parts)
     {
         Pipe pipe = createMergeTreeSequentialSource(
             MergeTreeSequentialSourceType::Merge,
             *global_ctx->data,
             global_ctx->storage_snapshot,
             part,
-            global_ctx->merging_column_names,
+            merging_column_names,
```
Comment (antaljanosbenjamin) [inline src/Storages/MergeTree/MergeTask.cpp:1009]:
```suggestion
    for (const auto & part : global_ctx->future_part->parts)
    {
        Pipe pipe = createMergeTreeSequentialSource(
            MergeTreeSequentialSourceType::Merge,
            *global_ctx->data,
            global_ctx->storage_snapshot,
            part,
            global_ctx->merging_columns.getNames(),
```

This way we won't copy `merging_column_names`.

File: tests/queries/0_stateless/03166_skip_indexes_vertical_merge_2.sql [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+--- merged: a, c, d; gathered: b, e, f
+WITH
+    (SELECT uuid FROM system.tables WHERE database = currentDatabase() AND table = 't_ind_merge_2') AS uuid,
+    extractAllGroupsVertical(message, 'containing (\\d+) columns \((\\d+) merged, (\\d+) gathered\)')[1] AS groups
+SELECT
+    groups[1] AS total,
+    groups[2] AS merged,
+    groups[3] AS gathered
+FROM system.text_log
+WHERE query_id = uuid || '::all_1_2_1' AND notEmpty(groups)
```
Comment (antaljanosbenjamin) [inline tests/queries/0_stateless/03166_skip_indexes_vertical_merge_2.sql:38]:
```suggestion
WHERE query_id = uuid || '::all_1_2_1' AND notEmpty(groups)
```
Isn't the `uuid` the uuid of the table? Do I miss something here? Maybe use `position(message, uuid) > 0`?

**Review by antaljanosbenjamin — APPROVED** [review]
I don't have too broad knowledge about the inner details of how vertical/horizontal merges are done, but based on my limited understanding all of the changes make sense and looks good.

File: tests/queries/0_stateless/03166_skip_indexes_vertical_merge_2.sql (hunk 2 above)
Comment (CurtizJ) [reply tests/queries/0_stateless/03166_skip_indexes_vertical_merge_2.sql]: Yes, it's uuid of table. We use `uuid::part_name` as query id for merges.

**Review by antaljanosbenjamin — APPROVED** [review]

**Pull request merged**